File: Lecture 1 - C/Problem Set 1/Mario-Less/mario.c
/*
    PROBLEM: MARIO (LESS COMFORTABLE) - STANDARD C VERSION

    Write a program that:
    - Prompts the user for a height between 1 and 8.
    - Prints a right-aligned pyramid.
    - No CS50 library used.
*/

#include <stdio.h>

// Function prototypes
int get_height(void);
void print_spaces(int space);
void print_rows(int row);

int main(void)
{
    int height = get_height();

    for (int i = 0; i < height; i++)
    {
        print_spaces(height - (i + 1));
        print_rows(i + 1);
    }

    return 0;
}

// Get height using scanf
int get_height(void)
{
    int height;

    do
    {
        printf("Height: ");
        scanf("%d", &height);
    }
    while (height < 1 || height > 8);

    return height;
}

// Print spaces
void print_spaces(int space)
{
    for (int i = 0; i < space; i++)
    {
        printf(" ");
    }
}

// Print hashes and newline
void print_rows(int row)
{
    for (int i = 0; i < row; i++)
    {
        printf("#");
    }

    printf("\n");
}

/*
---------------- ALGORITHM ----------------

1. Prompt user for integer height (1–8).
2. Repeat until valid height is entered.
3. For each row:
    a. Print decreasing spaces.
    b. Print increasing hashes.
    c. Move to next line.
4. End program.

Time Complexity: O(n²)
Space Complexity: O(1)
*/

File: Lecture 6 - Python/Problem Set 6/Mario - Less/mario.py
"""
PROBLEM: MARIO (LESS COMFORTABLE)

Write a program that:
1. Prompts the user for a height (integer between 1 and 8).
2. Prints a right-aligned half-pyramid of that height.
"""


# Keep asking user until valid input is entered
while True:
    try:
        # Take input and convert to integer
        height = int(input("Height: "))

        # Check if height is within valid range
        if height > 0 and height < 9:
            break
        else:
            print("Enter a number between 1 to 8")

    # If input cannot be converted to integer
    except ValueError:
        print("Enter a number between 1 to 8")


# Print the pyramid
for i in range(0, height, 1):

    # Print leading spaces
    for j in range(0, height - (i + 1), 1):
        print(" ", end="")

    # Print hashes
    for j in range(0, i + 1, 1):
        print("#", end="")

    # Move to next line after each row
    print("")


"""
---------------- ALGORITHM ----------------

1. Repeat forever:
    a. Try to take integer input from user.
    b. If height is between 1 and 8:
         break loop.
    c. Otherwise:
         print error message.
    d. If input is not an integer:
         catch ValueError and print error message.

2. For each row i from 0 to height - 1:
    a. Print (height - i - 1) spaces.
    b. Print (i + 1) hashes.
    c. Move to next line.

Time Complexity: O(n²)
Space Complexity: O(1)
"""